src/Workspaces/Remote/ServiceHub/Shared/RemoteEndPoint.cs

```diff
@@ -216,12 +217,19 @@ public async Task<T> InvokeAsync<T>(string targetName, IReadOnlyList<object?> ar
             }
         }
 
-        public static Task WriteDataToNamedPipeAsync<TData>(string pipeName, TData data, Func<ObjectWriter, TData, CancellationToken, Task> dataWriter, CancellationToken cancellationToken)
+        public static Task WriteDataToNamedPipeAsync<TData>(string pipeName, TData data, AssetStorages.Storage? storage, Func<ObjectWriter, TData, CancellationToken, Task> dataWriter, CancellationToken cancellationToken)
             => WriteDataToNamedPipeAsync(pipeName, data,
                 async (stream, data, cancellationToken) =>
                 {
-                    using var objectWriter = new ObjectWriter(stream, leaveOpen: true, canKeepObjectsAlive: false, cancellationToken);
+                    using var objectWriter = new ObjectWriter(stream, leaveOpen: true, canKeepObjectsAlive: storage is object, cancellationToken);
                     await dataWriter(objectWriter, data, cancellationToken).ConfigureAwait(false);
+                    if (storage is object)
+                    {
+                        foreach (var obj in objectWriter.TakeKeepAliveObjects())
+                        {
+                            storage.KeepAlive(obj);
+                        }
+                    }
```

Comment:
Could simplify to:

```suggestion
                        foreach (var obj in objectWriter.TakeKeepAliveObjects())
                        {
                            storage?.KeepAlive(obj);
                        }
```

In hte case that storage is null, TakeKeepAliveObjects would return an empty array anyways, so this would efectively be a no-op.

Reply:
so this seems to be the meat of the change.  and seein this now, i'm not a huge fan.  i think i would prefer that instead of having this array be filled that you copy out and then place into another array, you simply pass in either a collectoin or a callback to ObjectWriter stating: here's where you can put hte keepAlive objects.  So this code would just be:

```
new ObjectWriter(stream, leaveOpen: true, keepAliveCallback: storage?.KeepAlive, cancellationToken)
```

Reply:
> Could simplify to

I had it like this originally and didn't like the readability.

Reply:
> keepAliveCallback: storage?.KeepAlive

Am implementing this change. It has other benefits as well.

Reply:
➡️ Now implemented